### PR TITLE
Interval update - Parser for the Anderson dataset

### PR DESCRIPTION
**Pull request by DSuveges**
- The Anderson dataset is parsed according to the existing logic implemented by Ed. 
 - All operation is implemented in pyspark
 - Dataset is saved as parquet
 - LiftOver logic implemented
 - Gene annotation is also implemented. Currently the gene annotation is based on an ad hoc dataset, that will be replaced by the updated target index.

### CODE REVIEW

File: scripts/intervals_parser.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+    # Further parsers will come here...
+
+    # Saving data:
+    anderson_df.write.parquet(output)
+
+
+if __name__ == '__main__':
+    # Input datasets (adjust to your needs):
+    anderson_file = 'enhancer_tss_associations.bed'
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]:
```suggestion
    anderson_file = 'gs://genetics-portal-input/v2g_input/andersson2014/enhancer_tss_associations.bed'
```

File: scripts/intervals_parser.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    # LiftOver chain file (adjust to your needs):
+    chain_file = 'grch37_to_grch38.over.chain.gz'
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]:
```suggestion
    chain_file = 'gs://genetics-portal-input/v2g_input/grch37_to_grch38.over.chain.gz'
```

**Comment by ireneisdoomed** [comment]
@DSuveges I have updated the PR to accept remote GS locations. The bed file is read natively by Pandas and the chain file is handled with GCSFS. I also added a remote location for the output files. Please, could you take a look at the changes?

Now the PR can be submitted as a Spark job. These are the commands I've used to test the code more easily.
```
CLUSTER="il-v2g"
REGION="europe-west1"
PROJECT="open-targets-eu-dev"

# Cluster creation
gcloud dataproc clusters create ${CLUSTER} \
    --image-version=2.0 \
    --region=${REGION} \
    --metadata 'PIP_PACKAGES=pandas pyliftover' \
    --initialization-actions gs://goog-dataproc-initialization-actions-europe-west1/python/pip-install.sh                                    \
    --master-machine-type=n1-standard-32 \
    --master-boot-disk-size=100

# Job submission
gcloud dataproc jobs submit pyspark \
  --cluster=${CLUSTER} \
  --project=${PROJECT} \
  --region=${REGION} \
  scripts/intervals_parser.py
```

The output can be checked at `gs://genetics-portal-dev-staging/v2g/interval/dev/intervals-2022-07-13`

**Comment by DSuveges** [comment]
The output data schema:

```
root
 |-- chrom: string (nullable = true)
 |-- start: string (nullable = true)
 |-- end: string (nullable = true)
 |-- gene_id: string (nullable = true)
 |-- score: double (nullable = true)
 |-- dataset_name: string (nullable = true)
 |-- data_type: string (nullable = true)
 |-- experiment_type: string (nullable = true)
 |-- pmid: string (nullable = true)
 |-- bio_feature: string (nullable = true)
```

Table:
```
+-----+-------+-------+---------------+-----+-------------+---------+---------------+--------+-----------+
|chrom|  start|    end|        gene_id|score| dataset_name|data_type|experiment_type|    pmid|bio_feature|
+-----+-------+-------+---------------+-----+-------------+---------+---------------+--------+-----------+
|    1|1413305|1413316|ENSG00000215915|0.357|andersson2014| interval|        fantom5|24670763|  aggregate|
|    1|1413305|1413316|ENSG00000186891|0.659|andersson2014| interval|        fantom5|24670763|  aggregate|
|    1|1413305|1413316|ENSG00000205116|0.184|andersson2014| interval|        fantom5|24670763|  aggregate|
+-----+-------+-------+---------------+-----+-------------+---------+---------------+--------+-----------+
```
Current table:
```
+-----+-------+-------+---------------+-------------------+-----------+
|chrom|  start|    end|        gene_id|              score|bio_feature|
+-----+-------+-------+---------------+-------------------+-----------+
|    1| 922877| 923268|ENSG00000187634|0.16321600000000003|  aggregate|
|    1|1006412|1006755|ENSG00000188290|           0.034969|  aggregate|
|    1|1010390|1010654|ENSG00000160072|           0.031684|  aggregate|
+-----+-------+-------+---------------+-------------------+-----------+
```
This implementation produces a dataset with 54_194 lines vs the current dataset 55_941. The difference most likely due to the fact that the source data contains obsolete gene symbols. The existing pipelines uses gene annotation from 2018. I think at this point let these rows lost, but later we can go back and use the extra annotations in the target object to resolve these terms. (The currently hardwired gene table is an ad-hoc solution, without alternative gene names or obsolete synonyms)

File: configs/config.yaml [hunk 3]
```diff
@@ -10,4 +10,43 @@ genes: gs://genetics-portal-input/luts/19.06_gene_symbol_synonym_map.json
...
 # Interval Liftover
 chain_grch37_to_grch38: grch37_to_grch38.over.chain.gz
-max_len_dff: 100 # Max difference in length between old and new interval (in bases)
\ No newline at end of file
+max_len_dff: 100 # Max difference in length between old and new interval (in bases)
+
+
+intervals:
+  liftover_chain_file: gs://genetics-portal-input/v2g_input/grch37_to_grch38.over.chain
+  max_lenght_difference: 100
```
Comment (ireneisdoomed) [inline configs/config.yaml]:
```suggestion
  max_length_difference: 100
```

File: scripts/intervals_parser.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+    spark = (
+        pyspark.sql.SparkSession
+        .builder
+        .master("local[*]")
+        .getOrCreate()
+    )
+
+    chain_file = cfg.intervals.liftover_chain_file
+    max_difference = cfg.intervals.max_lenght_difference
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]:
```suggestion
    max_difference = cfg.intervals.max_length_difference
```

File: scripts/intervals_parser.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+import pandas as pd
+
+import hydra
+
+from pyliftover import LiftOver
+
+from pyspark.sql import dataframe
+import pyspark.sql
+import pyspark.sql.types as t
+import pyspark.sql.functions as f
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]: So far we've usually imported the Functions module as `F`. Should we adopt it as a new convention? Same applies to the types.

File: scripts/intervals_parser.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+from pyspark.sql import SparkSession
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]:
```suggestion
from pyspark.sql import dataframe, SparkSession
```

File: scripts/intervals_parser.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
+from datetime import date
+import gcsfs
+import pandas as pd
+
+import hydra
+
+from pyliftover import LiftOver
+
+from pyspark.sql import dataframe
+import pyspark.sql
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]:
```suggestion
```

File: scripts/intervals_parser.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            .filter((f.col('mapped').isNotNull()) & (f.size(f.col('mapped')) == 1))
+
+            # Extracting mapped corrdinates:
+            .withColumn('mapped_' + chrom_name, f.col('mapped')[0][0])
+            .withColumn('mapped_' + pos_name, f.col('mapped')[0][1])
+
+            # Drop rows that mapped to the other chromosomes:
+            .filter(f.col('mapped_' + chrom_name) == f.concat(f.lit('chr'),f.col(chrom_name)))
+
+            # Dropping unused coluns:
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]:
```suggestion
            # Dropping unused columns:
```

File: scripts/intervals_parser.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        # Lift over start coordinates:
+        start_df = df.select(chrom_col, start_col).distinct()
+        start_df = self.convert_coordinates(start_df, chrom_col, start_col).withColumnRenamed('mapped_pos', 'mapped_' + start_col)
+
+        # Lift over end coordinates:
+        end_df = df.select(chrom_col, end_col).distinct()
+        end_df = self.convert_coordinates(end_df, chrom_col, end_col).withColumnRenamed('mapped_pos', 'mapped_' + end_col)
+
+        # Join dataframe with mappings:
+        return (
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]: Instead of creating 2 DFs to add each of the mapped positions and then join these back to the main df, I think it is simpler if you operate on the main df all the time like this:

```
return (
    df
    .transform(lambda df: self.convert_coordinates(df, chrom_col, start_col))
    .withColumnRenamed('mapped_pos', 'mapped_' + start_col)
    .transform(lambda df: self.convert_coordinates(df, chrom_col, end_col))
    .withColumnRenamed('mapped_pos', 'mapped_' + end_col)

    # ...
)
```

File: scripts/intervals_parser.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+            df
+            .join(start_df, on=[chrom_col, start_col], how='left')
+            .join(end_df, on=[chrom_col, end_col], how='left')
+
+            # Select only rows where the start is smaller than the end:
+            .filter(
+                # Drop rows with no mappings:
+                f.col('mapped_start').isNotNull() & f.col('mapped_end').isNotNull()
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]:
```suggestion
                f.col('mapped_' + start_col).isNotNull() & f.col('mapped_' + end_col).isNotNull()
```

File: scripts/ensembl_gtf_to_tss_bed.py [hunk 11]
```diff
@@ -31,8 +31,19 @@ def main():
         df = df.loc[to_keep, :]
 
     # Convert rows to bed format
-    bed = df.apply(gtf_row_to_bed, axis=1)
+    bed =(
+        df.head(10000)
```
Comment (ireneisdoomed) [inline scripts/ensembl_gtf_to_tss_bed.py]:
```suggestion
df
```

File: scripts/intervals_parser.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+from pyliftover import LiftOver
+
+from pyspark.sql import dataframe
+import pyspark.sql
+import pyspark.sql.types as t
+import pyspark.sql.functions as f
+from pyspark.sql import SparkSession
+
+
+class LiftOverSpark:
```
Comment (ireneisdoomed) [inline scripts/intervals_parser.py]: I would move this class to an utils module.

**Review by ireneisdoomed — COMMENTED** [review]
Thanks a lot Daniel, I've really enjoyed reviewing this PR. It looks fantastic and I think it sets a standard for how we should work on the rest of the pipelines.

I have some overall comments or opinions on top of the other comments:
1. How is the module `ensembl_gtf_to_tss` used in the pipeline? Is the content of the bed file it produces expected to be found in the gene index?

2. I see the value in moving all the environment variables to the config, so that we can grasp what those are in a single file. However, for values that will remain constant like the `dataset_name`, `data_type`, `experiment_type` and `bio_feature`, I am concerned that we eventually end up with really long configs. In the end for each release, we want to go over and edit every variable that needs to be updated. The easier this process is, the better.

3. Do you think you can also delete the files in the repo that have gotten obsolete with your changes? Things like andersson2014_fantom5.Snakefile, grch37_to_grch38.over.chain.gz, interval_master.gcloud.sh...

File: configs/config.yaml (hunk 3 above)
Comment (DSuveges) [reply configs/config.yaml]: Thanks for spotting!

File: scripts/ensembl_gtf_to_tss_bed.py (hunk 11 above)
Comment (DSuveges) [reply scripts/ensembl_gtf_to_tss_bed.py]: I actually accidentally saved this modification, I did not have an intention to make any change in this module. It was more of an exploration to find out what the hell this script is doing. I don't like in this pipeline that Ed didn't use column names instead column indices. Very hard to follow.

File: scripts/intervals_parser.py (hunk 5 above)
Comment (DSuveges) [reply scripts/intervals_parser.py]: OK.

File: scripts/intervals_parser.py (hunk 12 above)
Comment (DSuveges) [reply scripts/intervals_parser.py]: Yes. Most likely I'm going to split the parsers out as well into separate modules. It' getting way too long.

File: scripts/intervals_parser.py (hunk 9 above)
Comment (DSuveges) [reply scripts/intervals_parser.py]: Oh boy, I'm not sure about that. There's some optimization in that step: when you say:
```Python
df.select(chrom_col, start_col).distinct()
```
You reduce the number of positions you need to lift over. By omitting this step (although the code would be nicer), you potentially need to map the same position over and over again.

File: scripts/intervals_parser.py (hunk 10 above)
Comment (DSuveges) [reply scripts/intervals_parser.py]: Thanks for finding this! This bug caused me some headache. (the function didn't work if you wanted to lift over other columns but `start` and `end`)

**Comment by DSuveges** [comment]
@ireneisdoomed Thanks for the comments and the kind words! :blush: I'm doing my best to make the process smooth and consistent and as easy to understand as possible. Currently it's quite hard to follow the code some times.

Answering you questions:
1. Currently the `ensembl_gtf_to_tss` pipeline is used to generate a bed file with the gene ids, symbols and TSS. This needs to be phased out as the target index from the platform should bring this in.
2. Yes I was also considering that, and will do it. At the same time parsers for each dataset will be in a separate file.
3. Yes, I'll remove the obsoleted files.

File: scripts/intervals_parser.py (hunk 9 above)
Comment (ireneisdoomed) [reply scripts/intervals_parser.py]: Ok, this is a very good point that I had not considered. Let's leave it this way then. Thanks ^^

File: configs/config.yaml [hunk 13]
```diff
@@ -10,4 +10,43 @@ genes: gs://genetics-portal-input/luts/19.06_gene_symbol_synonym_map.json
...
+intervals:
+  liftover_chain_file: gs://genetics-portal-input/v2g_input/grch37_to_grch38.over.chain
+  max_length_difference: 100
+  output: gs://genetics-portal-dev-data/dev_intervals
+  gene_index: gs://genetics-portal-input/v2g_input/geneset.parquet
+  proximity_limit: 5.341328e05
+  # An atlas of active enhancers across human cell types and tissues
+  anderson:
+    data_file: gs://genetics-portal-input/v2g_input/andersson2014/enhancer_tss_associations.bed
+    dataset_name: andersson2014
```
Comment (ireneisdoomed) [inline configs/config.yaml:25]: If you agree with my comment # 2, these parameters need to be removed from the config. Otherwise it is fine.

File: intervals_parser.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        .select('chrom', 'start', 'end', 'gene_symbol', 'score')
+
+        # Drop rows with non-canonical chromosomes:
+        .filter(f.col('chrom').isin([str(x) for x in range(1, 23)] + ['X', 'Y', 'MT']))
+
+        # For each region/gene, keep only one row with the highest score:
+        .groupBy('chrom', 'start', 'end', 'gene_symbol')
+        .agg(f.max('score').alias('score'))
+
+        .orderBy('chrom', 'start')
```
Comment (ireneisdoomed) [inline intervals_parser.py]: What is the utility of having the data ordered?

File: intervals_parser.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    max_difference = cfg.intervals.max_length_difference
+    proximity_limit = 2 * float(cfg.intervals.proximity_limit)
+
+    # Open and process gene file:
+    gene_index = spark.read.parquet(cfg.intervals.gene_index).persist()
+
+    # Initialize liftover object:
+    lift = LiftOverSpark(chain_file, max_difference)
+
+    # Parsing the anderson file:
```
Comment (ireneisdoomed) [inline intervals_parser.py]: I'd suggest defining an array containing the different datasets like here https://github.com/opentargets/evidence_datasource_parsers/blob/215adb032afd1f23acac2e72973b2a6844301d90/modules/GeneBurden.py#L29

File: intervals_parser.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    anderson_df = parse_anderson(cfg.intervals.anderson, gene_index, lift, proximity_limit)
```
Comment (ireneisdoomed) [inline intervals_parser.py]: I think this needs updating, now the function to parse Andersson data should be coming from `andersson2014_parser.py`.

File: modules/andersson2014_parser.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
+
```
Comment (ireneisdoomed) [inline modules/andersson2014_parser.py:3]: You need to import Spark

File: modules/Liftover.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
+import gcsfs
+
+from pyliftover import LiftOver
+
+import pyspark.sql.types as T
+import pyspark.sql.functions as F
```
Comment (ireneisdoomed) [inline modules/Liftover.py:11]: If Pyspark functions are imported as F, the instances need to be updated.

I have a preference for using an upper case as it is the way I most commonly see them referenced. It is a pain when you want to reproduce some snippet in the REPL and I have an error because of differences in how the API is imported.
Databricks import them using upper case in [their docs](https://docs.gcp.databricks.com/spark/latest/dataframes-datasets/introduction-to-dataframes-python.html), also from the[ Data Analysis with Python and Pyspark book](https://books.google.pt/books?id=4ytfEAAAQBAJ&pg=PA53&lpg=PA53&dq=from+pyspark.sql+import+functions+as+F+or+as+f&source=bl&ots=24AtuuBozV&sig=ACfU3U2K-N-de_fBhZXRlAB2hPqL1NpjMw&hl=es&sa=X&ved=2ahUKEwi83OKxuoL5AhXvm_0HHaE5C1YQ6AF6BAg4EAM#v=onepage&q=from%20pyspark.sql%20import%20functions%20as%20F%20or%20as%20f&f=false)

Again, no hard feelings, I just prefer to make a single decision and adapt the rest of the code to that.

**Review by ireneisdoomed — COMMENTED** [review]
Thank you so much for addressing the comments! There are currently some very minor things to finish, like to import the logic of the parsing from the new independent Andersson module.

File: intervals_parser.py (hunk 14 above)
Comment (DSuveges) [reply intervals_parser.py]: Once the data is collected it will be partitioned by chrom. Other than that... I might as well drop.
